src/ingest/src/main/scala/osmesa/functions/osm/package.scala

```diff
@@ -0,0 +1,294 @@
+              case None => throw new AssemblyException("Unable to connect segments.")
+            }
+        }, rings)
+    }
+  }
+
+  private def connectSegments(segments: List[Line]): List[Polygon] = {
+    connectSegments(segments.map(_.jtsGeom.getCoordinates): List[Array[Coordinate]]).map(_.map(Point.jtsCoord2Point)).map(Polygon(_))
+  }
+
+  @tailrec
+  private def dissolveRings(rings: List[Polygon], dissolvedOuters: List[Polygon] = List.empty[Polygon], dissolvedInners: List[Polygon] = List.empty[Polygon]): (List[Polygon], List[Polygon]) = {
+    rings match {
+      case Nil => (dissolvedOuters, dissolvedInners)
+      case h :: t =>
+        t.filter(r => h.touches(r)) match {
+          case touching if touching.isEmpty => dissolveRings(t.filterNot(r => h.touches(r)), dissolvedOuters :+ Polygon(h.exterior), dissolvedInners ++ h.holes.map(Polygon(_)))
+          case touching =>
+            val dissolved = touching.foldLeft(List(h)) {
+              case (rs, r2) =>
+                rs.flatMap { r =>
+                  r.union(r2).toGeometry match {
+                    case Some(p: Polygon) => List(p)
+                    case Some(mp: MultiPolygon) => mp.polygons
+                    case _ => throw new AssemblyException("Union failed.")
+                  }
+                }
+            }
+
+            val remaining = t.filterNot(r => h.touches(r))
+            val retryRings = dissolved.filter(d => remaining.exists(r => r.touches(d)))
+            val newRings = dissolved.filter(d => !remaining.exists(r => r.touches(d)))
+
+            dissolveRings(retryRings ++ remaining, dissolvedOuters ++ newRings.map(_.exterior).map(Polygon(_)), dissolvedInners ++ newRings.flatMap(_.holes).map(Polygon(_)))
+        }
+    }
+  }
+
+  def buildMultiPolygon(id: Long, version: Long, timestamp: Timestamp, types: Seq[Byte], roles: Seq[String], wkbs: Seq[Array[Byte]]): Array[Byte] = {
+    if (types.zip(wkbs).exists { case (t, g) => t == ProcessOSM.WAY_TYPE && Option(g).isEmpty }) {
+      // bail early if null values are present where they should exist (members w/ type=way)
+      logger.debug(s"Incomplete relation: $id @ $version ($timestamp)")
+      return null
+    }
+
+    val geoms = wkbs.map(Option(_).map(_.readWKB) match {
+      case Some(geom: Polygon) => geom.as[Polygon].map(_.exterior)
+      case Some(geom) => geom.as[Line]
+      case None => None
+    })
+
+    val members: Seq[(String, Line)] = roles.zip(geoms)
+      .filter(_._2.isDefined)
+      .map(x => (x._1, x._2.get))
+
+    val (completeOuters, completeInners, completeUnknowns, partialOuters, partialInners, partialUnknowns) = members.foldLeft((List.empty[Polygon], List.empty[Polygon], List.empty[Polygon], List.empty[Line], List.empty[Line], List.empty[Line])) {
+      case ((co, ci, cu, po, pi, pu), (role, geom: Line)) =>
+        Option(geom) match {
+          case Some(line) =>
```

Comment:
Confused with this - you wrap the `geom` in an Option so it'll always be `Some`?

Reply:
Uh... Not sure why I did this either. Addressed in b689988.

Reply:
That could account for some garbage.